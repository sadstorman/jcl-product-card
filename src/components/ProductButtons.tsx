import { ProductContext } from './ProductCard';
import React, { CSSProperties, useCallback, useContext } from 'react';
import styles from "../styles/styles.module.css";

export interface Props {
    className?: string;
    style?: CSSProperties
}

export const ProductButtons = ({ className, style }: Props) => {

    // Extraer mas count del context
    const { counter, increaseBy, maxCount } = useContext(ProductContext)

    // isMaxReached = useCallback, dependencias [counter, maxCount] 
    // TRUE si el count === maxCount
    // FaLSE si es distinto
    const isMaxReached = useCallback(
        () => !!maxCount && counter === maxCount,
        [counter, maxCount]
    )
    

    return (
        <div className={`${styles.buttonsContainer} ${className} `}
            style={style}
        >
            <button className={styles.buttonMinus} onClick={() => { increaseBy(-1) }} >
                -
            </button>

            <div className={styles.countLabel} >
                {counter}
            </div>

            <button className={`${styles.buttonAdd} ${(isMaxReached()) ? styles.disabled : ''}`} onClick={() => { increaseBy(+1) }}>
                +
            </button>
        </div>
    )
}