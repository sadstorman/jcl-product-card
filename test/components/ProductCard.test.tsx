import React from 'react';
import renderer from 'react-test-renderer'
import { ProductCard } from '../../src/components';
import { producto1 } from './data/products';

const {act} = renderer

describe('Product Card', () => {

    test('Debe mostrar el componente correctamente', () => {
        const wrapper = renderer.create(

            <ProductCard product={producto1}>
                {
                    () => (
                        <h1> Product Card</h1>
                    )
                }
            </ProductCard>
        )

        expect(wrapper.toJSON()).toMatchSnapshot()
    })

    test('Debe de incrementar el contador', () => {
        const wrapper = renderer.create(

            <ProductCard product={producto1}>
                {
                    ({count, increaseBy}) => (
                        <>
                        <h1> Product Card</h1>
                        <span>{count}</span>
                        <button onClick={ ()=> increaseBy(+1)}> +1</button>
                        </>
                    )
                }
            </ProductCard>
        );

        let tree = wrapper.toJSON()
        expect(tree).toMatchSnapshot();

        act(() => {
            (tree as any).children[2].props.onClick()
        })

        tree = wrapper.toJSON()

        expect((tree as any).children[1].children[0]).toBe('1')
    })
})