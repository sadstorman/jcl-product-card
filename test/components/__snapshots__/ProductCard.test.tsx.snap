// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Card Debe de incrementar el contador 1`] = `
<div
  className="productCard undefined"
>
  <h1>
     Product Card
  </h1>
  <span>
    0
  </span>
  <button
    onClick={[Function]}
  >
     +1
  </button>
</div>
`;

exports[`Product Card Debe mostrar el componente correctamente 1`] = `
<div
  className="productCard undefined"
>
  <h1>
     Product Card
  </h1>
</div>
`;
